src/db.c

```diff
@@ -192,24 +230,46 @@ robj *lookupKeyWriteOrReply(client *c, robj *key, robj *reply) {
  * if the key already exists, otherwise, it can fall back to dbOverwite. */
 static void dbAddInternal(redisDb *db, robj *key, robj *val, int update_if_existing) {
     dictEntry *existing;
-    dictEntry *de = dictAddRaw(db->dict, key->ptr, &existing);
+    int slot = getKeySlot(key->ptr);
+    dict *d = db->dict[slot];
+    dictEntry *de = dictAddRaw(d, key->ptr, &existing);
     if (update_if_existing && existing) {
         dbSetValue(db, key, val, 1, existing);
         return;
     }
     serverAssertWithInfo(NULL, key, de != NULL);
-    dictSetKey(db->dict, de, sdsdup(key->ptr));
+    dictSetKey(d, de, sdsdup(key->ptr));
     initObjectLRUOrLFU(val);
-    dictSetVal(db->dict, de, val);
+    dictSetVal(d, de, val);
+    db->key_count++;
+    cumulativeKeyCountAdd(db, slot, 1);
     signalKeyAsReady(db, key, val->type);
-    if (server.cluster_enabled) slotToKeyAddEntry(de, db);
     notifyKeyspaceEvent(NOTIFY_NEW,"new",key,db->id);
 }
 
 void dbAdd(redisDb *db, robj *key, robj *val) {
     dbAddInternal(db, key, val, 0);
 }
 
+/* Returns key's hash slot when cluster mode is enabled, or 0 when disabled.
+ * The only difference between this function and getKeySlot, is that it's not using cached key slot from the current_client
+ * and always calculates CRC hash.
+ * This is useful when slot needs to be calculated for a key that user didn't request for, such as in case of eviction. */
+int calculateKeySlot(sds key) {
+    return server.cluster_enabled ? keyHashSlot(key, (int) sdslen(key)) : 0;
+}
+
+/* Return slot-specific dictionary for key based on key's hash slot in CME, or 0 in CMD.*/
+int getKeySlot(sds key) {
+    /* This is performance optimization, that uses pre-set slot id from the current command,
+     * in order to avoid calculation of the key hash. Code paths that are using keys, that can be from different slots,
+     * MUST unset current client's slot value before calling any db functions, otherwise wrong dictionary can be used. */
+    if (server.current_client && server.current_client->slot >= 0) {
+        return server.current_client->slot;
+    }
```

Comment:
i think there's a problem here with scripts that declared SCRIPT_ALLOW_CROSS_SLOT.

this also brings me to a realization, that i'm not sure we have any way to detect bugs in this mechanism.
i.e. we don't have tests that can detect keys being placed in the wrong dict.
and we don't have any assertions when the binary index tree is decremented below 0.

Reply:
@hpatro this one is now resolved with the change you did in `tests/unit/cluster/scripting.tcl`, right?
please add a comment in these checks you added there describing their purpose.